File: src/main/java/de/fluxparticle/animation/object/ElementNode.java
package de.fluxparticle.animation.object;

import de.fluxparticle.animation.signal.Signal;
import javafx.beans.binding.DoubleExpression;

public abstract class ElementNode {

	protected final DoubleExpression time;

	private final Signal<Double> opacity;
	public Signal<Double> opacitySignal() {
		return opacity;
	}

	public abstract <R> R accept(ElementNodeVisitor<R> visitor);

	public ElementNode(DoubleExpression time) {
		this.time = time;
		this.opacity = new Signal<>(time, this, "opacity", 0.0);
	}

}

File: src/main/kotlin/de/fluxparticle/animation/object/ElementNode.kt
package de.fluxparticle.animation.`object`

import de.fluxparticle.animation.signal.Signal
import javafx.beans.binding.DoubleExpression

abstract class ElementNode(protected val time: DoubleExpression) {

    private val opacity: Signal<Double>
    fun opacitySignal(): Signal<Double> {
        return opacity
    }

    abstract fun <R> accept(visitor: ElementNodeVisitor<R>): R

    init {
        this.opacity = Signal(time, this, "opacity", 0.0)
    }

}
